### ISSUE TITLE
Wallaby + webpack + babel + rewire

### ISSUE THREAD

**Issue by codyhatch** [wallabyjs/public#157]
I'd like to use rewire with webpack and babel.  I have [this babel plugin](https://github.com/speedskater/babel-plugin-rewire) working with my Karma setup, but I can't seem to figure out how to get it working with Wallaby.

My wallaby.js file:

```
var wallabyWebpack = require('wallaby-webpack');
var babel = require('babel');

module.exports = function (wallaby) {

    var babelCompiler = wallaby.compilers.babel({babel});

    var babelPreprocessor = file => babel.transform(file.content, {plugins: ["rewire"]});

    var webpackPostprocessor = wallabyWebpack({
        resolve: {
            extensions: ['', '.js', '.jsx']
        },
        module: {
            loaders: [
                {test: /\.less$/, loader: "style-loader!css-loader!less-loader"}
            ]
        }
    });

    return {
        files: [
            {pattern: 'node_modules/react-tools/src/test/phantomjs-shims.js', instrument: false},
            {pattern: 'node_modules/sinon/pkg/sinon.js', instrument: false},
            {pattern: 'node_modules/chai/chai.js', instrument: false},
            {pattern: 'src/**/*.js*', load: false},
            {pattern: 'src/**/*.less', load: false},
            {pattern: 'src/**/__tests__/*.js', ignore: true}
        ],

        tests: [
            {pattern: 'src/**/__tests__/*.js', load: false}
        ],

        compilers: {
            '**/*.js*': babelCompiler
        },

        preprocessor: {
            '**/*.js*': babelPreprocessor
        },

        postprocessor: webpackPostprocessor,

        testFramework: "mocha@2.0.1",

        bootstrap: function () {
            window.BASE_URL = "";
            window.API_SESSION_KEY = "123456";
            window.HEADOFFICE_ID = 1;
            window.expect = chai.expect;
            window.__moduleBundler.loadTests();
        }
    };
};
```

Which gave the error `'_SearchForm2['default'].__Rewire__'`.

I tried removing the preprocessor and adding babel as a loader for the webpack postprocessor:

```
    var webpackPostprocessor = wallabyWebpack({
        resolve: {
            extensions: ['', '.js', '.jsx']
        },
        module: {
            loaders: [
                { test: /\.jsx?$/, loader: 'babel-loader?plugins=rewire', exclude: /node_modules/ },
                {test: /\.less$/, loader: "style-loader!css-loader!less-loader"}
            ]
        }
    });
```

But that yielded the same error.  I also tried passing the plugin as an option to the babel compiler:

```
    var babelCompiler = wallaby.compilers.babel({
        babel: babel,
        // babel options
        stage: 0,
        plugins: ['rewire']
    });
```

Which yields the error `Failed to run compilers on src/common/utils/sorting.js, ReferenceError: The plugin "wallaby.js" collides with another of the same name`, so I guess you can't use babel plugins that way.

For reference, the karma.conf.js file which works looks like this:

```
var webpack = require('webpack');

module.exports = function(config) {
    config.set({
        browsers: ['PhantomJS'],
        singleRun: false,
        frameworks: ['mocha', 'chai-sinon'],
        files: [
            './globals.js',
            {
                pattern: 'tests.webpack.js',
                watched: false
            }
        ],
        preprocessors: {
            'tests.webpack.js': 'webpack'
        },

        reporters: [ 'spec' ], //report results in this format

        webpack: {
            devtool: false,
            module: {
                loaders: [
                    { test: /\.jsx?$/, loader: 'babel-loader?plugins=rewire', exclude: /node_modules/ },
                    { test: /\.less$/, loader: "style-loader!css-loader!less-loader" }
                ]
            },
            watch: true
        },
        webpackServer: {
            noInfo: true
        }
    });
};
```

Any help would be appreciated.  For the record, I'm trying to use rewire in a test like this:

```
        SearchForm.__Rewire__('RouteHandler', React.createClass({
            render: () => {
                return <div className="FakeRouteHandler"/>;
            }
        }));
```

(Basically, I have a React component that tried to render the RouteHandler component from react-router.  I need to replace it with a place holder, since the RouteHandler won't work in my test.)

**Comment by ArtemGovorov** [comment]
You may just add the rewire plugin to babel options when creating the compiler and will not need a preprocessor. I also had to add `Object.assign` shim to the `bootstrap` function because `rewire` generates some code that uses it and PhantomJs 1.9.8 doesn't support it. You may place it to a separate file if you like and reference it similar to `node_modules/react-tools/src/test/phantomjs-shims.js`, as opposed to having it in the wallaby config. You may also consider using [PhantomJs 2](http://wallabyjs.com/docs/integration/phantomjs2.html) and get rid of both shims.

So the working config is:

``` JavaScript
var wallabyWebpack = require('wallaby-webpack');
var babel = require('babel');

module.exports = function (wallaby) {

  var babelCompiler = wallaby.compilers.babel({babel, plugins: ['rewire']});

  var webpackPostprocessor = wallabyWebpack({
    resolve: {
      extensions: ['', '.js', '.jsx']
    },
    module: {
      loaders: [
        {test: /\.less$/, loader: "style-loader!css-loader!less-loader"}
      ]
    }
  });

  return {
    files: [
      {pattern: 'node_modules/react-tools/src/test/phantomjs-shims.js', instrument: false},
      {pattern: 'node_modules/sinon/pkg/sinon.js', instrument: false},
      {pattern: 'node_modules/chai/chai.js', instrument: false},
      {pattern: 'src/**/*.js*', load: false},
      {pattern: 'src/**/*.less', load: false},
      {pattern: 'src/**/__tests__/*.js', ignore: true}
    ],

    tests: [
      {pattern: 'src/**/__tests__/*.js', load: false}
    ],

    compilers: {
      '**/*.js*': babelCompiler
    },

    postprocessor: webpackPostprocessor,

    testFramework: "mocha@2.0.1",

    bootstrap: function () {
      if (!Object.assign) {
        Object.defineProperty(Object, 'assign', {
          enumerable: false,
          configurable: true,
          writable: true,
          value: function(target) {
            'use strict';
            if (target === undefined || target === null) {
              throw new TypeError('Cannot convert first argument to object');
            }

            var to = Object(target);
            for (var i = 1; i < arguments.length; i++) {
              var nextSource = arguments[i];
              if (nextSource === undefined || nextSource === null) {
                continue;
              }
              nextSource = Object(nextSource);

              var keysArray = Object.keys(Object(nextSource));
              for (var nextIndex = 0, len = keysArray.length; nextIndex < len; nextIndex++) {
                var nextKey = keysArray[nextIndex];
                var desc = Object.getOwnPropertyDescriptor(nextSource, nextKey);
                if (desc !== undefined && desc.enumerable) {
                  to[nextKey] = nextSource[nextKey];
                }
              }
            }
            return to;
          }
        });
      }
      window.BASE_URL = "";
      window.API_SESSION_KEY = "123456";
      window.HEADOFFICE_ID = 1;
      window.expect = chai.expect;
      window.__moduleBundler.loadTests();
    }
  };
};
```

Let me know if it works for you.

**Comment by codyhatch** [comment]
Thanks for the fast reply!  However, Using that config, I get the error in wallaby:

```
Failed to run compilers on src/pages/HeadOfficeOrders/components/__tests__/SearchFormSpec.js, ReferenceError: The plugin "wallaby.js" collides with another of the same name
  at PluginManager.validate (/Users/cody/Coding/Mobi2Go/admin-x/node_modules/babel-core/lib/babel/transformation/file/plugin-manager.js:82:13)
  at PluginManager.add (/Users/cody/Coding/Mobi2Go/admin-x/node_modules/babel-core/lib/babel/transformation/file/plugin-manager.js:135:10)
  at File.buildTransformers (/Users/cody/Coding/Mobi2Go/admin-x/node_modules/babel-core/lib/babel/transformation/file/index.js:290:21)
  at new File (/Users/cody/Coding/Mobi2Go/admin-x/node_modules/babel-core/lib/babel/transformation/file/index.js:148:10)
  at TransformerPipeline.transform (/Users/cody/Coding/Mobi2Go/admin-x/node_modules/babel-core/lib/babel/transformation/transformer-pipeline.js:96:16)

```

**Comment by ArtemGovorov** [comment]
Ok, I must be missing something then. Here is the repo with the structure with your config that works for me: https://github.com/wallabyjs/wallaby-rewire. Could you please clone it, run `npm install` and see if it works for you?

**Comment by ArtemGovorov** [comment]
Ah, never mind, reproduce the error you've mentioned. It starts occurring when I edit the code.

**Comment by ArtemGovorov** [comment]
The issue is fixed in core 1.0.65. The update will be installed automatically within the next 20 minutes after wallaby.js restart (or restart WebStorm to force the update and then start wallaby.js).

**Comment by codyhatch** [comment]
Thanks, issue is resolved.

(Great support, by the way.)
